Firestore/core/src/firebase/firestore/api/document_reference.h

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_API_DOCUMENT_REFERENCE_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_API_DOCUMENT_REFERENCE_H_
+
+#if !defined(__OBJC__)
+#error "This header only supports Objective-C++"
+#endif  // !defined(__OBJC__)
+
+#import <Foundation/Foundation.h>
+
+#include <string>
+#include <utility>
+
+#import "FIRDocumentReference.h"
+#import "FIRFirestoreSource.h"
+#import "FIRListenerRegistration.h"
+
+#include "Firestore/core/src/firebase/firestore/model/document_key.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@class FIRCollectionReference;
+@class FIRFirestore;
+@class FSTListenOptions;
+
+namespace firebase {
+namespace firestore {
+namespace api {
+
+class DocumentReference {
+ public:
+  using Completion = void (^)(NSError* _Nullable error) _Nullable;
+  using DocumentCompletion = void (^)(FIRDocumentSnapshot* _Nullable document,
+                                      NSError* _Nullable error) _Nullable;
+
+  DocumentReference() = default;
+  DocumentReference(FIRFirestore* firestore, model::DocumentKey document_key)
+      : firestore_{firestore}, key_{std::move(document_key)} {
+  }
+
+  size_t Hash() const;
+
+  FIRFirestore* firestore() const {
+    return firestore_;
+  }
+  const model::DocumentKey& key() const {
+    return key_;
+  }
+
+  std::string document_id() const;
+
+  FIRCollectionReference* parent() const;
+
+  std::string path() const;
+
+  FIRCollectionReference* collection(const std::string& collection_path) const;
```

Comment:
Strictly speaking this isn't a "property" in the way that the style guide calls for properties to be named this way. `GetCollection`? This would mirror `GetDocument` below, but the call semantics are different. `GetCollectionReference`?

Reply:
Related: http://go/c-readability-advice#virtual-accessor-names

Reply:
FWIW, I've only been using snake_case for direct accessors (eg getter1). But I don't mind getter2 at all.

Another 'rule-of-thumb' I've been using is that I only snake_case things I've also inlined in the header. (But I have a lot of sympathy for moving them into the cc file regardless of that to keep the clutter down. shrug)

Reply:
Direct accessors and mutators are clear cases where the snake-case names apply, but it's not required.

The name is a promise that the cost is similar to direct member access, so I think `getter2` is appropriate above. I agree these should be inlined in the header.

A getter calling a non-getter should be an immediate code smell.

Anything more than ~one line in the method body seems like it would be suspicious.

Reply:
Ok, I renamed `path` back to `Path`, and `parent` back to `Parent`, because neither is a getter.